Mathlib/GroupTheory/SpecificGroups/Dihedral.lean

```diff
@@ -256,5 +257,51 @@ lemma card_conjClasses_odd (hn : Odd n) :
   rw [←Nat.mul_div_mul_left _ 2 hn.pos, ← card_commute_odd hn, mul_comm,
     card_comm_eq_card_conjClasses_mul_card, nat_card, Nat.mul_div_left _ (mul_pos two_pos hn.pos)]
 
+section SmallOrder
+
+open scoped Classical
+
+variable {G : Type} [Group G] [Fintype G]
+
+/-- A non-cyclic group of order 4 is isomorphic to the dihedral group of order 4. -/
+noncomputable def mulEquivOfCardEqFourAndNotIsCyclic (h1 : Fintype.card G = 4) (h2 : ¬IsCyclic G) :
+    G ≃* DihedralGroup 2 :=
```

Comment:
I think it might be slightly shorter to do this constructively, by passing in two non-equal non-one elements of `G`, and identifying them with `sr 0` and `sr 1`

Reply:
I tried that and I'm not sure. It seems difficult to do this constructively, given that G is opaque. Am I missing something?

Reply:
No, it's entirely possible that this suggestion is just a bad idea after all. My thinking was that if you have `(sr0 sr1 : G)`, you can get `r1 : G := sr0 * sr1`, and use `h1` to case-bash on the three elements. The reason I suggest that is because it looks like your proof of bijectivity is effectively doing the same case-bash anyway.

Reply:
Pattern matching doesn't work like that I'm afraid, you have to chain `if then else`s

Reply:
Then I think I misunderstood what you meant by "constructive". I'll try again.